ToDoList/src/ToDoList.Contracts/DependencyInjection/IContainer.cs

```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace ToDoList.Contracts.DependencyInjection
+{
+    public interface IContainer
+    {
+        object GetContainer();
```

Comment:
Using object here was the only way how to separate implementation of DI (IUnityContainer) from other parts of the application (like ToDo.Contracts -> IDependencyRegister, where IContract is used).
I have tried to go with IContainer<T> ... or T GetContainer<T>, but every time I was forced to add IUnityContainer somewhere in the other project (like I mentioned - in ToDo.Contracts). Therefore using object seems like a right solution as I can determine in UnityDependencyResolver if the returned container has the correct type.

Reply:
That's because you are trying to wrap the `unity container` only partially - DI has two parts:
* _registration_ – that's what `IContainer` already does
* _resolution_ – that's what you need the `GetContainer`* method for since `IContainer` does not provide any way to resolve registered types

We later played around with the idea of separating these two parts into separate interfaces with Miro. However, in general, DI frameworks will usually provide you with the container exactly to allow you resolving on it (e.g. [`InjectionFactory` in `Unity`](https://msdn.microsoft.com/en-us/library/ff661868(v=pandp.51).aspx)). On the other hand, implementation of `IDependencyResolver` is the only place that needs the container for its resolving capabilities in your case and needs to resolve types in pretty loosely typed way (it expects an `object` to be returned while generic would ensure that the returned instance is an instance of the expected type or its successor as soon as during build). 

I don't believe you can go the very same way as Miro, since your solutions are approached from bit different angles. You might still create an interface allowing resolution (basically just 2 methods – `ResolveType` and `ResolveTypes`  – that would propagate towards wrapped `IUnityContainer`) and add another method to the `ContainerFactory`. You would have to make it create only one instance of the container at all times (there is the `Lazy` class to the rescue of course). This leaves two ways the container is used separate, however, it expects you to know that there is such factory that contains such method that returns such interface.

You could also introduce the interface described above and change the `GetContainer`* method's signature to return it (method would return `this`). This, however, allows anybody to obtain the resolution-enabled interface at any time (especially in one of `IDependencyRegister` implementations) – there is then no reason to separate the two interfaces. Mind you, we are still speaking about the very same instance of the container wrapper what so ever, so it would be better to simply extend `IContainer` with `ResolveType` and `ResolveTypes` methods (and "stay true and open" about the `IContainer`).

\* – the name is not clear now, you should've used something like `GetInnerContainer` or something that would explain it is not the same container (or its copy) that is being returned from the method.